src/host/_stream.cpp

```diff
@@ -1065,36 +1053,58 @@ using Microsoft::Console::VirtualTerminal::StateMachine;
 // - S_OK if successful.
 // - S_OK if we need to wait (check if ppWaiter is not nullptr).
 // - Or a suitable HRESULT code for math/string/memory failures.
-[[nodiscard]] HRESULT WriteConsoleWImplHelper(IConsoleOutputObject& context,
-                                              const std::wstring_view buffer,
-                                              size_t& read,
-                                              bool requiresVtQuirk,
-                                              std::unique_ptr<WriteData>& waiter) noexcept
+static [[nodiscard]] HRESULT WriteConsoleWImplHelper(
+    IConsoleOutputObject& context,
+    const std::wstring_view& buffer,
+    size_t& read,
+    bool requiresVtQuirk,
+    std::unique_ptr<WriteData>& waiter) noexcept
 {
-    try
+    // Set out variables in case we exit early.
+    read = 0;
+    waiter.reset();
+
+    RETURN_HR_IF(E_UNEXPECTED, buffer.size() > SIZE_T_MAX / sizeof(wchar_t));
+
+    // Convert characters to bytes to give to DoWriteConsole.
+    auto data = const_cast<wchar_t*>(buffer.data());
+    auto remaining = buffer.size();
+    NTSTATUS Status = S_OK;
+
+    while (remaining != 0)
     {
-        // Set out variables in case we exit early.
-        read = 0;
-        waiter.reset();
+        auto chunkLength = std::min<size_t>(remaining * sizeof(wchar_t), 4 * 1024 * 1024);

```

Comment:
@DHowett This is the input chunking I meant btw. It's being chunked inside conhost and thus shouldn't affect the semblance of us being synchronous from the perspective of the client.